### PR TITLE
Add TLS support to POP3

### PR DESCRIPTION
**Pull request by cfpwastaken**
This PR adds TLS support to POP3.
- You can run both POP3 and POP3S at the same time (but both can be toggled individually)
- You can specify a custom cert

New config options:
- enable_pop3 (Default true): Can be used to disable the unencrypted pop3 server
- enable_pop3s (Default false): Can be used to enable the encrypted pop3 server
- tls_key: A path to the private key
- tls_cert: A path to the certificate

POP3S is not enabled by default, because by default it obv. doesnt ship with a cert.

### CODE REVIEW

**Comment by cfpwastaken** [comment]
Both the POP3 and SMTP specs say that you can also start TLS-ing on the unencrypted port by using the STARTTLS command, but I dont think that would be easy to implement + using the TLS port is better.
Should we add an error message to tell the client to just use the tls port when trying to use the STARTTLS command?

**Comment by j0code** [comment]
@cfpwastaken Sounds good, but only if the spec provides a way to do that. Otherwise, return a 502.

**Comment by cfpwastaken** [comment]
spec doesnt allow, so we keep what we have rn.

File: src/main.ts [hunk 1]
```diff
@@ -22,4 +22,5 @@ export const sql = new Sequelize({
 // })
 
 export const smtpserver = new SMTPServer(getConfig("smtp_port", 25)) // Port 25 for regular SMTP, 465 for SMTPS
-const pop3server = new POP3Server(getConfig("pop3_port", 110)) // Port 110 for regular POP3, 995 for POP3S
+const pop3server = getConfig("enable_pop3", true) ? new POP3Server(getConfig("pop3_port", 110), false) : null // Port 110 for regular POP3, 995 for POP3S
+const pop3sserver = getConfig("enable_pop3s", false) ? new POP3Server(getConfig("pop3s_port", 995), true) : null // Port 110 for regular POP3, 995 for POP3S
```
Comment (j0code) [inline src/main.ts:26]:
```suggestion
if (getConfig("enable_pop3", true)) new POP3Server(getConfig("pop3_port", 110), false) // Port 110 for regular POP3, 995 for POP3S
if (getConfig("enable_pop3s", false)) new POP3Server(getConfig("pop3s_port", 995), true) // Port 110 for regular POP3, 995 for POP3S
```

File: src/pop3/POP3Server.ts [hunk 2]
```diff
@@ -1,23 +1,30 @@
...
 	server: net.Server
+	useTLS: boolean
 
-	constructor(port: number) {
-		this.server = net.createServer()
+	constructor(port: number, useTLS: boolean) {
+		this.useTLS = useTLS
+		this.server = useTLS ? tls.createServer({
+			key: readFileSync(getConfig("tls_key", "cert/privkey.pem")),
+			cert: readFileSync(getConfig("tls_cert", "cert/fullchain.pem"))
```
Comment (j0code) [inline src/pop3/POP3Server.ts:22]: 1. use fs/promises
2. error handling - what if the paths provided aren't valid, the file does not exist or it is inaccessible?

**Review by j0code — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
 node_modules
 config.json
-mails
\ No newline at end of file
+mails
+cert
```
[the remark this answers is not in the document]
Comment (cfpwastaken) [reply .gitignore:4]: good idea

File: src/pop3/POP3Server.ts (hunk 2 above)
Comment (cfpwastaken) [reply src/pop3/POP3Server.ts:22]: ok gonna look

File: src/main.ts (hunk 1 above)
Comment (j0code) [reply src/main.ts:26]: yes. fairly confident to the point of certainty.

**Comment by cfpwastaken** [comment]
I literally made an empty commit and it unlocked

File: src/pop3/POP3Server.ts [hunk 4]
```diff
@@ -1,24 +1,31 @@
 import net from "net"
-import { readdirSync, readFileSync } from "fs"
+import tls from "tls"
+import { existsSync, readdirSync, readFileSync } from "fs"
```
Comment (j0code) [inline src/pop3/POP3Server.ts:3]: remove unused import
```suggestion

```

File: src/pop3/POP3Server.ts [hunk 5]
```diff
@@ -1,24 +1,31 @@
...
 import User from "../models/User.js"
 import { createHash } from "node:crypto"
-import { readFile } from "fs/promises"
 import Mail from "../models/Mail.js"
 import Logger from "../Logger.js"
 import getConfig from "../config.js"
+import { readFile } from "fs/promises"
```
Comment (j0code) [inline src/pop3/POP3Server.ts:9]:
```suggestion

```
this one too

**Review by j0code — CHANGES_REQUESTED** [review]

File: src/pop3/POP3Server.ts (hunk 5 above)
Comment (cfpwastaken) [reply src/pop3/POP3Server.ts:9]: This one is used in line 90 and line 96

Comment (j0code) [reply src/pop3/POP3Server.ts:8]: right, didn't see that.

**Review by j0code — APPROVED** [review]

**Pull request merged**